mesmerize_viz/_common.py

```diff
@@ -0,0 +1,185 @@
+            reset the timepoint in the ImageWidget when changing items/rows
+
+        input_movie_kwargs: dict, optional
+            kwargs passed to get_input_movie()
+
+        image_widget_kwargs: dict, optional
+            kwargs passed to ImageWidget
+
+        Returns
+        -------
+        ImageWidget
+            fastplotlib.ImageWidget visualization
+        """
+
+        if input_movie_kwargs is None:
+            input_movie_kwargs = dict()
+
+        if image_widget_kwargs is None:
+            image_widget_kwargs = dict()
+
+        # the ones which are [t, x, y] images that ImageWidget can manage by itself
+        self.standard_mappings = standard_mappings
+
+        # data arrays directly passed to image widget
+        data_arrays_iw = self._parse_data(
+            data=data,
+            data_mapping=data_mapping,
+            input_movie_kwargs=input_movie_kwargs
+        )
+
+        # default kwargs unless user has specified
+        default_iw_kwargs = {
+            "window_funcs": {"t": (np.mean, 11)},
+            "vmin_vmax_sliders": True,
+            "cmap": "gnuplot2"
+        }
+
+        image_widget_kwargs = {
+            **default_iw_kwargs,
+            **image_widget_kwargs
+        }
+
+        self.image_widget = ImageWidget(
+            data=data_arrays_iw,
+            names=data,
+            **image_widget_kwargs
+        )
+
+        for a, n in zip(data_arrays_iw, data):
+            if isinstance(a, ZeroArray):
+                # rename the existing graphic
+                self.image_widget.plot[n].graphics[0].name = "zero-array-ignore"
+                # get the real data
+                func = data_mapping[n]
+                real_data = func()
+                # create graphic with the real data, this will not be managed by ImageWidget
+                self.image_widget.plot[n].add_image(real_data, name="image", cmap="gnuplot2")
+
+        self.reset_timepoint_on_change = reset_timepoint_on_change
+
+    def _parse_data(
+            self,
+            data: List[str],
+            data_mapping: dict,
+            input_movie_kwargs: dict,
+    ) -> List[Union[np.ndarray, ZeroArray]]:
+        """
+        Parse data string keys into actual arrays using the data_mapping for the current row.
+        Returns list of arrays that ImageWidget can display and manage.
+        """
+        # data arrays directly passed to image widget
+        data_arrays_iw = list()
+
+        for d in data:
+            if d in self.standard_mappings:
+                func = data_mapping[d]
+
+                if d == "input":
+                    a = func(**input_movie_kwargs)
+                else:
+                    a = func()
+
+                data_arrays_iw.append(a)
+
+            else:
+                # make a placeholder array to keep imagewidget happy
+                # hacky but this is the best way for now
+                zero_array = ZeroArray(ndim=data_arrays_iw[0].ndim)
```

Comment:
oh, I see now...zero array for placeholder

Reply:
yes because ImageWidget needs arrays of all same dims to be happy, so just give it a "ZeroArray" which just returns a 0 regardless of how you index it (it just subclasses LazyArray)